### ISSUE TITLE
standalone cudacpp code generation fails after integration of multichannel

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#473]
I have debugged several issues in #419 and #472 for PR #465 about multichannel, using the --mad generation option.

I am now here: https://github.com/valassi/madgraph4gpu/commit/c3908de807409666b9e24902af821a2194000831

I now tried to produce standalone cudacpp directories, but I get
```
Command "import /data/avalassi/GPU2020/MG5aMC/3.1.1_lo_vectorization/CODEGEN_cudacpp_ee_mumu.mg" interrupted in sub-command:
"output standalone_cudacpp CODEGEN_cudacpp_ee_mumu" with error:
AssertionError : 
Please report this bug on https://bugs.launchpad.net/mg5amcnlo
More information is found in 'MG5_debug'.
Please attach this file to your report.

...

  File "/data/avalassi/GPU2020/MG5aMC/3.1.1_lo_vectorization/madgraph/iolibs/export_cpp.py", line 915, in get_process_f
unction_definitions
    self.get_all_sigmaKin_lines(color_amplitudes,
  File "/data/avalassi/GPU2020/MG5aMC/3.1.1_lo_vectorization/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py", line 1068, in
 get_all_sigmaKin_lines
    assert self.include_multi_channel # do we really need to generate code where this is not true?
AssertionError
```

**Comment by valassi** [comment]
I have now simply removed the assertion. Code generation does go through.

But I have discovered that the generated code is not the same with and without mad...

In https://github.com/valassi/madgraph4gpu/commit/4f518515804f9f621b65c1216f3fe668b0f05b5f
```
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/ ee_mumu.auto/SubProcesses/P1_Sigma_sm_epem_mupmum/
...
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/CPPProcess.cc ee_mumu.auto/SubProcesses/P1_Sigma_sm_epem_mupmum/CPPProcess.cc
226,227d225
<       if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
237,238d234
<       if( channelId == 2 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/CPPProcess.h ee_mumu.auto/SubProcesses/P1_Sigma_sm_epem_mupmum/CPPProcess.h
68c68
<     //static const int wrows = 5; // mgOnGpu::nw6;
---
>     //static const int wrows = 6; // mgOnGpu::nw6;
...
```

I find the difference in wrows especially bizarre?!

**Comment by valassi** [comment]
Tracing back where this comes from.

This is from my very abrupt backport to code generation back in https://github.com/valassi/madgraph4gpu/commit/d3f82900d8b465e79dfd573c1bcbb85ff503f574

```
git show d3f82900d8 CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py 
commit d3f82900d8b465e79dfd573c1bcbb85ff503f574
Author: Andrea Valassi <andrea.valassi@cern.ch>
Date:   Mon May 23 00:45:20 2022 +0200

    [mch] backport CPPProcess.cc

diff --git a/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py b/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py
index 374ea857..8a0620f3 100644
--- a/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py
+++ b/epochX/cudacpp/CODEGEN/PLUGIN/CUDACPP_SA_OUTPUT/model_handling.py
@@ -1065,21 +1065,21 @@ class PLUGIN_OneProcessExporter(PLUGIN_export_cpp.OneProcessExporterGPU):
         """Get sigmaKin_process for all subprocesses for gCPPProcess.cu"""
         ret_lines = []
         if self.single_helicities:
-            if self.include_multi_channel:
-                info = {'multi_channel': self.multichannel_var}  
-            else:
-                info = {'multi_channel': ''}
+            assert self.include_multi_channel # do we really need to generate code where this is not true?
             ret_lines.append("""
   // Evaluate |M|^2 for each subprocess
   // NB: calculate_wavefunctions ADDS |M|^2 for a given ihel to the running sum of |M|^2 over helicities for the given event(s)
+  // (similarly, it also ADDS the numerator and denominator for a given ihel to their running sums over helicities)
   __device__ INLINE void /* clang-format off */
   calculate_wavefunctions( int ihel,
-                           const fptype* allmomenta,   // input: momenta[nevt*npar*4]
-                           const fptype* allcouplings, // input: couplings[nevt*ndcoup*2]
-                           fptype* allMEs              // output: allMEs[nevt], |M|^2 running_sum_over_helicities""")
-            ret_lines.append( '%(multi_channel)s' % info )
-            ret_lines.append("""#ifndef __CUDACC__
-                           , const int nevt            // input: #events (for cuda: nevt == ndim == gpublocks*gputhreads)
+                           const fptype* allmomenta,    // input: momenta[nevt*npar*4]
+                           const fptype* allcouplings,  // input: couplings[nevt*ndcoup*2]
+                           fptype* allMEs,              // output: allMEs[nevt], |M|^2 running_sum_over_helicities
+                           fptype* allNumerators,       // output: multichannel numerators[nevt], running_sum_over_helicities
+                           fptype* allDenominators,     // output: multichannel denominators[nevt], running_sum_over_helicities
+                           const unsigned int channelId // input: multichannel channel id (1 to #diagrams); 0 to disable channel enhancement
+#ifndef __CUDACC__
+                           , const int nevt             // input: #events (for cuda: nevt == ndim == gpublocks*gputhreads)
 #endif
                            )
...
```

Essentially, the original implementation of code generation by Olivier was foreseeing different code generated with and without multichannel. I just tried to take a shortcut and assume that I would always (and would always be able to!) egenerate multichannel

**Comment by valassi** [comment]
Hi @oliviermattelaer I am giving up, sorry I need your help...

Repeating the above, my question is essentially the following: is it possible for me to generate the same exact cudacpp code in standalone mode and in madevent+2ndExporter mode? Before the integration of multichannel this was possible. Since I integrated multichannel, I have the impression that this is no longer possible. 

I would really like to have exactly the same code base (quite simply, because building ggttggg takes 2 hours for my tests, and I prefer to do it only once, in the madevent+2ndExporter mode, rather than having to maintain also the standalone mode).

I have tried several patches but there seems to be some fundamental problem, where for instance the line `self.include_multi_channel = config_map` in `generate_process_files_madevent` can realy only go through if one uses the madevent plus second exporter mode. I just tried another patch where I tried to get 'diagram_maps' from the helas matrix element, but this fails badly. I just do not understand well enough how these python instances interact.

Summary of the summary:
- is there hope that I can have the same multichannel code even in standalone cudacpp mode, and how?
- or should I give up and (like you do in standalone_gpu) produce different code in the two modes, standalone_cudacpp and madevent + cudacpp?

Last comment on having multichannel code in standalone mode: of course it is not really used as we do not have a multichannel integrator, just the simple rambo. But the whole point of the standalone cudacpp is to test the same code we can then send to madevent. So it is very easy for me to give a channel number as an option to the standalone check.exe and gcheck.exe, and instead of the full ME calculate the single-diagram-enghanced ME for a given channel. This is actaly what I was planning in #466. If this #473 cannot be fixed, then I will abandon that one too, but it's a pity.

Thanks! Talk to you next Tuesday, have a good weekend
Andrea

**Comment by oliviermattelaer** [comment]
Well,

Since I do not like ifdef, I typically like to handle those are different output.
But it should be possible with a lot of ifdef 
(since at least the API are not compatible)

But if it takes you that long for that process, maybe I should take a look at why (I do have guess and if I'm right i do have have flag to speed-up the generation --with likely a slightly less optimized code--

**Comment by valassi** [comment]
Hi Olivier, thanks. 

Hm actually here I did not even think we needed an ifdef? I actually do not want an ifdef here (I would like to be able to build the same cudacpp code ony once, whether multichannel is on ie channelId>0, or multichannel is off ie channelId=0). I was thinking of a runtime choice, not a compile time choice.

I mean, in practice the only thing that seems to change in the generated code for me is, quoting above
```
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/CPPProcess.cc ee_mumu.auto/SubProcesses/P1_Sigma_sm_epem_mupmum/CPPProcess.cc
226,227d225
<       if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
237,238d234
<       if( channelId == 2 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
```
Then I guess there are other places further down where some actions for multichannel are only done if channelId is not 0.

The idea here was that you can pass channelId=0 and then numerators and denominators are just not used. So in a way you can 'disable multichannel' by passing 0 as an argument (which is something I want to be able to do anyway also in madevent mode), you do not need to completely remove all of the code.

Similarly, my idea was that you can pass channelId>0 also if you are in standalone mode (NOT in madevent mode), so that you test what MEs are given as output.

Anyway if you prefer to have two different code bases just let me know (there may be good reasons for it), then I will follow more closely what you are doing in the standalone gpu case. If instead you agree we could give it a try to have a single code base, with runtime choice in cudacppp dependening on channelId=0 or >0, then I think I need some help with the python code generation machinery.

Thanks!
a

**Comment by valassi** [comment]
I think that the issue goes through here
https://github.com/mg5amcnlo/mg5amcnlo/blob/b09f37181ae4670730dbd1d4e202add02adfcd21/madgraph/iolibs/export_v4.py#L1226
```
    def get_multi_channel_dictionary(diagrams, config_map): 
        """diagrams should be from matrix_element.get('diagrams')"""


        config_to_diag_dict = {}
        if config_map:
            # In this case, we need to sum up all amplitudes that have
            # identical topologies, as given by the config_map (which
            # gives the topology/config for each of the diagrams
            # Combine the diagrams with identical topologies
            for idiag, diag in enumerate(diagrams):
                if config_map[idiag] == 0:
                    continue
                try:
                    config_to_diag_dict[config_map[idiag]].append(idiag)
                except KeyError:
                    config_to_diag_dict[config_map[idiag]] = [idiag]
        else:
            # Get minimum legs in a vertex
            vert_list = [max(diag.get_vertex_leg_numbers()) for diag in \
                diagrams if diag.get_vertex_leg_numbers()!=[]]
            minvert = min(vert_list) if vert_list!=[] else 0

            for idiag, diag in enumerate(diagrams):
                # Ignore any diagrams with 4-particle vertices.
                if diag.get_vertex_leg_numbers()!=[] and max(diag.get_vertex_leg_numbers()) > minvert:
                    continue
                config_to_diag_dict[config_map[idiag]] = [idiag]

        return  config_to_diag_dict
```

Essentially the question is, is there a way in standalone_cudacpp mode to provide a config_map to this method?

**Comment by oliviermattelaer** [comment]
a real one no...
now we can pass a fake config_map {i:i}
but it really depends of what you want to do...



On 7 Jun 2022, at 15:28, Andrea Valassi ***@***.******@***.***>> wrote:



I think that the issue goes through here
https://github.com/mg5amcnlo/mg5amcnlo/blob/b09f37181ae4670730dbd1d4e202add02adfcd21/madgraph/iolibs/export_v4.py#L1226

    def get_multi_channel_dictionary(diagrams, config_map):
        """diagrams should be from matrix_element.get('diagrams')"""


        config_to_diag_dict = {}
        if config_map:
            # In this case, we need to sum up all amplitudes that have
            # identical topologies, as given by the config_map (which
            # gives the topology/config for each of the diagrams
            # Combine the diagrams with identical topologies
            for idiag, diag in enumerate(diagrams):
                if config_map[idiag] == 0:
                    continue
                try:
                    config_to_diag_dict[config_map[idiag]].append(idiag)
                except KeyError:
                    config_to_diag_dict[config_map[idiag]] = [idiag]
        else:
            # Get minimum legs in a vertex
            vert_list = [max(diag.get_vertex_leg_numbers()) for diag in \
                diagrams if diag.get_vertex_leg_numbers()!=[]]
            minvert = min(vert_list) if vert_list!=[] else 0

            for idiag, diag in enumerate(diagrams):
                # Ignore any diagrams with 4-particle vertices.
                if diag.get_vertex_leg_numbers()!=[] and max(diag.get_vertex_leg_numbers()) > minvert:
                    continue
                config_to_diag_dict[config_map[idiag]] = [idiag]

        return  config_to_diag_dict


Essentially the question is, is there a way in standalone_cudacpp mode to provide a config_map to this method?

**Comment by valassi** [comment]
Thanks Olivier! Hm what I want to do is essentially only being able to add these lines in code generation also in standalone_cudacpp mode. Now I am only able to do it in madevent + 2nd exporter mode
```
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/CPPProcess.cc ee_mumu.auto/SubProcesses/P1_Sigma_sm_epem_mupmum/CPPProcess.cc
226,227d225
<       if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
237,238d234
<       if( channelId == 2 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
```

**Comment by valassi** [comment]
Ok this is now fixed.

As discussed, this now leads to two codebases: standalone_cudacpp and madevent+cudacpp. These are the only differences:

First
```
diff -r ee_mumu.mad/src/ ee_mumu/src/
...
diff -r ee_mumu.mad/src/mgOnGpuConfig.h ee_mumu/src/mgOnGpuConfig.h
6c6
< #define MGONGPU_SUPPORTS_MULTICHANNEL 1
---
> #undef MGONGPU_SUPPORTS_MULTICHANNEL
```


Second
```
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/ ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/
...
diff -r ee_mumu.mad/SubProcesses/P1_ll_ll/CPPProcess.cc ee_mumu/SubProcesses/P1_Sigma_sm_epem_mupmum/CPPProcess.cc
242,243c242
<       if( channelId == 1 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
---
>       // Here the code base generated with multichannel support updates numerators_sv and denominators_sv (#473)
255,256c254
<       if( channelId == 2 ) numerators_sv += cxabs2( amp_sv[0] );
<       if( channelId != 0 ) denominators_sv += cxabs2( amp_sv[0] );
---
>       // Here the code base generated with multichannel support updates numerators_sv and denominators_sv (#473)
```

Note that in the end, for simplicity in code generation, I did (unfortunately) go to ifdefs. So most of the code (numerators and denominators in the API) is there in both code bases, but protected by MGONGPU_SUPPORTS_MULTICHANNEL. The only part of the code that is really different (and that causes the need to have two code bases and MGONGPU_SUPPORTS_MULTICHANNEL in the first place) is those two lines above updating numerators_sv and denominators_sv.

This will be fixed by PR #465. Closing.
